livekit-agents/livekit/agents/llm/function_context.py

```diff
@@ -14,94 +14,156 @@
 
-        # retrieve ai functions & metadata
-        for _, member in inspect.getmembers(self, predicate=inspect.ismethod):
-            if not hasattr(member, METADATA_ATTR):
-                continue
 
-            metadata: AIFncMetadata = getattr(member, METADATA_ATTR)
-            if metadata.name in self._fncs:
-                raise ValueError(f"Duplicate function name: {metadata.name}")
+@dataclass(frozen=True)
+class FunctionInfo:
+    name: str
+    description: str
+    auto_retry: bool
+    callable: Callable
+    arguments: dict[str, FunctionArgInfo]
 
-            sig = inspect.signature(member)
-            type_hints = typing.get_type_hints(member)  # Annotated[T, ...] -> T
-            args = dict()
 
-            for name, param in sig.parameters.items():
-                if param.kind not in (
-                    inspect.Parameter.POSITIONAL_OR_KEYWORD,
-                    inspect.Parameter.KEYWORD_ONLY,
-                ):
-                    raise ValueError(
-                        f"unsupported parameter kind inside ai_callable: {param.kind}"
-                    )
+@dataclass(frozen=True)
+class CalledFunction:
+    info: FunctionInfo
+    arguments: dict[str, Any]
+    task: asyncio.Task
+    result: Any | None = None
+    exception: BaseException | None = None
 
-                if param.annotation is inspect.Parameter.empty:
-                    raise ValueError(f"missing type annotation for parameter {name}")
 
-                th = type_hints[name]
+def ai_callable(
+    *,
+    name: str | None = None,
+    description: str | _UseDocMarker | None = None,
+    auto_retry: bool = False,
+) -> Callable:
+    def deco(f):
+        _set_metadata(f, name=name, desc=description, auto_retry=auto_retry)
+        return f
 
-                if not is_type_supported(th):
-                    raise ValueError(f"unsupported type {th} for parameter {name}")
+    return deco
 
-                default = param.default
 
-                type_info = _find_param_type_info(param.annotation)
-                desc = type_info.desc if type_info else ""
+class FunctionContext:
+    def __init__(self) -> None:
+        self._fncs = dict[str, FunctionInfo]()
 
-                args[name] = AIFncArg(
-                    name=name,
-                    desc=desc,
-                    type=th,
-                    default=default,
+        for _, member in inspect.getmembers(self, predicate=inspect.ismethod):
+            if hasattr(member, METADATA_ATTR):
+                self._register_ai_function(member)
+
+    def ai_callable(
+        self,
+        *,
+        name: str | None = None,
+        description: str | _UseDocMarker | None = None,
+        auto_retry: bool = True,
+    ) -> Callable:
+        def deco(f):
+            _set_metadata(f, name=name, desc=description, auto_retry=auto_retry)
+            self._register_ai_function(f)
+
+        return deco
+
+    def _register_ai_function(self, fnc: Callable) -> None:
```

Comment:
Should we have a public version of this? Should we have an unregister method? Maybe out of scope for this pr